src/transport_performance/utils/raster.py

```diff
@@ -0,0 +1,167 @@
+    input_dir: str,
+    output_dir: str,
+    output_filename: str,
+    subset_regex: str = None,
+) -> dict:
+    """Merge raster files together.
+
+    Takes an input directory and merges all `.tif` files within it. Writes
+    merged raster to file.
+
+    Parameters
+    ----------
+    input_dir : str
+        Directory containing input raster files.
+    output_dir : str
+        Directory to write output, merged raster file.
+    output_filename : str
+        Filename of merged raster file (.tif extension required).
+    subset_regex : str, optional
+        Subset any raster files in the input directory using a regex, by
+        default None which means no subsetting will occur.
+
+    Returns
+    -------
+    bounds : dict
+        A dictionary summarising the boundaries of all input rasters and the
+        merged output. The "inputs" key is a list of the respective input
+        boundaries. The "output" key is a list containing the bounds of the
+        merged result. Useful to checking consistency of merged output.
+
+    Raises
+    ------
+    FileNotFoundError
+        In scenarios where `input_dir` does not exist and if there are no .tif
+        detectable within `input_dir`.
+
+    Notes
+    -----
+    1. This function does not provide any consistency checking of inputs and
+    merged outputs (e.g., checking for overlapping inputs and 'gaps' in the
+    merged outputs). This is primarily because merging is nuanced and numerous
+    in the potential ways inputs can be merged. For this reason, it is down to
+    the function's user to ensure the merged output is consistent with the
+    respective inputs. To this end, the bounds dict is returned to allow user
+    consistency testing.
+
+    2. The default rioxarry behaviours are assumed when merging inputs, i.e.,
+    the `CRS`, resolution and `nodata` values will be taken from the first
+    input DataArray. See [1]_ for more details.
+
+    References
+    ----------
+    .. [1] https://corteva.github.io/rioxarray/html/rioxarray.html#rioxarray.m
+    erge.merge_arrays
+
+    """
+    # defend against case where the provided input dir does not exist
+    if not os.path.exists(input_dir):
+        raise FileNotFoundError(f"{input_dir} can not be found")
+
+    # get tif files in directory, ensure some exist and select subset via regex
+    tif_filepaths = glob.glob(f"{input_dir}/*.tif")
+    if len(tif_filepaths) == 0:
+        raise FileNotFoundError(f"No `*.tif` files found in {input_dir}")
+
+    # apply regex and ensure tif files exist after applying it
+    # raise a unique FileNotFoundError to aid debugging
+    if subset_regex is not None:
+        tif_filepaths = [
+            fpath for fpath in tif_filepaths if re.search(subset_regex, fpath)
+        ]
+    if len(tif_filepaths) == 0:
+        raise FileNotFoundError(
+            f"No `*.tif` files found in {input_dir} after applying regex "
+            f"'{subset_regex}'."
+        )
```

Comment:
Test this behaviour with a check that no tif files matching pattern found. Probably need a tmpdir with some empty tifs for this, though no filenames should match the regex pattern being searched for.